### PR TITLE
fix: Handle multiple nil values on unique indexed fields

### LINKED ISSUE #2174: Violating a unique index with a nil value returns unhelpful error, or panics
For example, the below will return a `The given field does not exist. Name: age` error instead of an index violation error when attempting to create `Shahzad`.  Worse, if the index is created after all three documents (instead of before `Shahzad`), the code panics with `Unclosed iterator at time of Txn.Discard`:

Error:
```
                       testUtils.SchemaUpdate{
				Schema: `
					type User {
						name: String 
						age: Int
					}
				`,
			},
			testUtils.CreateDoc{
				CollectionID: 0,
				Doc: `
					{
						"name":	"John",
						"age":	21
					}`,
			},
			testUtils.CreateDoc{
				CollectionID: 0,
				Doc: `
					{
						"name":	"Andy"
					}`,
			},
			testUtils.CreateIndex{
				CollectionID: 0,
				FieldName:    "age",
				Unique:       true,
			},
			testUtils.CreateDoc{
				CollectionID: 0,
				Doc: `
					{
						"name":	"Shahzad"
					}`,
			},
```
Panic:
```
                       testUtils.SchemaUpdate{
				Schema: `
					type User {
						name: String 
						age: Int
					}
				`,
			},
			testUtils.CreateDoc{
				CollectionID: 0,
				Doc: `
					{
						"name":	"John",
						"age":	21
					}`,
			},
			testUtils.CreateDoc{
				CollectionID: 0,
				Doc: `
					{
						"name":	"Andy"
					}`,
			},
			testUtils.CreateDoc{
				CollectionID: 0,
				Doc: `
					{
						"name":	"Shahzad"
					}`,
			},
			testUtils.CreateIndex{
				CollectionID: 0,
				FieldName:    "age",
				Unique:       true,
			},
```

### PR DESCRIPTION
**Pull request by islamaliev**
## Relevant issue(s)

Resolves #2174 #2175

## Description

Forbid `nil` values for fields that are unique-indexed.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Integration and unit tests

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

**Comment by AndrewSisley** [comment]
This does feel like the wrong approach.   I think users should still be allowed nil values in indexed fields, especially considering that the GQL types *are* nillable (`Int` vs `INT!`).

The SQL standards for this allow multiple nil values in a column with a unique constraint.  Microsoft SQL Server naturally ignores this, and only permits a single nil value.

https://stackoverflow.com/questions/767657/how-do-i-create-a-unique-constraint-that-also-allows-nulls (check top answer).

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
This does feel like the wrong approach.   I think users should still be allowed nil values in indexed fields, especially considering that the GQL types *are* nillable (`Int` vs `Int!`).

The SQL standards for this allow multiple nil values in a column with a unique constraint.  Microsoft SQL Server naturally ignores this, and only permits a single nil value.

https://stackoverflow.com/questions/767657/how-do-i-create-a-unique-constraint-that-also-allows-nulls (check top answer).

File: tests/integration/index/create_unique_test.go [hunk 1]
```diff
@@ -186,3 +186,90 @@ func TestUniqueIndexCreate_IfFieldValuesAreUnique_Succeed(t *testing.T) {
...
+				Unique:        true,
+				ExpectedError: db.NewErrCanNotIndexNonUniqueField("bae-caba9876-89aa-5bcf-bc1c-387a52499b27", "age", nil).Error(),
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestUniqueIndexCreate_UponAddingDocWithExistingNilValue_ReturnError(t *testing.T) {
```
Comment (AndrewSisley) [inline tests/integration/index/create_unique_test.go:236]: todo: Can you please add a test that shows a single document with a nil indexed value succeeding?  This way if this (multiple nil values) test is deleted (e.g. when switching to allowing multiple nils) we do not lose coverage of the single nil happy-case.

todo: Can you please add a test with a filter eq nil?

todo: Can you please add some tests for composite indexes with nil and non-nil values?

File: db/index.go [hunk 2]
```diff
@@ -290,10 +290,16 @@ func (i *collectionUniqueIndex) newUniqueIndexError(
 	doc *client.Document,
 ) error {
 	fieldVal, err := doc.GetValue(i.fieldDesc.Name)
+	var val any
 	if err != nil {
-		return err
+		if !errors.Is(err, client.ErrFieldNotExist) {
```
Comment (AndrewSisley) [inline db/index.go:295]: todo: Can you please document why this if exists, it will not be obvious when reading the code at a later date.

File: db/errors.go [hunk 3]
```diff
@@ -86,7 +86,7 @@ const (
 	errExpectedJSONArray                  string = "expected JSON array"
 	errOneOneAlreadyLinked                string = "target document is already linked to another document"
 	errIndexDoesNotMatchName              string = "the index used does not match the given name"
-	errCanNotIndexNonUniqueField          string = "can not create doc that violates unique index"
+	errCanNotIndexNonUniqueField          string = "can not index a doc's field that violates unique index"
```
Comment (AndrewSisley) [inline db/errors.go:89]: praise: Thanks for solving this one :)

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
Looks good, and I'm glad not much production code was required.  I do think more tests are required though.

File: tests/integration/index/create_unique_test.go (hunk 1 above)
Comment (AndrewSisley) [reply tests/integration/index/create_unique_test.go:236]: Could you please add it as an iteration test please?  It will make it much easier to discover, and it will be able to take advantage of all our integration test related perks (such as mutation and client types, change detection, etc)

Comment (AndrewSisley) [reply tests/integration/index/create_unique_test.go:236]: > the current test already includes a successful addition of a document with a nil value:

Yes I saw that, but an independent successful test would be useful. If this (multiple nil values) test is deleted (e.g. when switching to allowing multiple nils) we do not lose coverage of the single nil happy-case.

Comment (islamaliev) [reply tests/integration/index/create_unique_test.go:261]: added an explicit test

**Review by AndrewSisley — APPROVED** [review]
LGTM :)
